common/interactive-object/server.mjs

```diff
@@ -0,0 +1,262 @@
+import {MapCoord} from '../maplib/map.mjs';
+
+/**
+ * TODO
+ * @param {Object} initialPosition - TODO
+ * @return {Boolean}
+ */
+function initialPositionValidation(initialPosition) {
+  // TODO
+  return true;
+}
+
+/**
+ * TODO
+ * @param {Object} fsm - TODO
+ * @return {Boolean}
+ */
+function fsmValidation(fsm) {
+  // TODO
+
+  // for (const attr of ['initialState', 'states']) {
+  //   if (!(attr in fsm)) {
+  //     console.error(`missing attribute '${attr}' of FSM`);
+  //     return false;
+  //   }
+  // }
+
+  // for (const [name, state] of Object.entries(fsm.states)) {
+  //   for (const attr of ['func', 'kwargs']) {
+  //     if (!(attr in state)) {
+  //       console.error(`missing attribute '${attr}' of state '${name}'`);
+  //       return false;
+  //     }
+  //   }
+  //   // TODO: security concern of using eval()
+  //   if (typeof eval(state.func) !== 'function') {
+  //     console.error(`'${state.func}' is not a valid function.`);
+  //     return false;
+  //   }
+
+  //   state.name = name;
+  // }
+
+  // return true;
+  return true;
+}
+
+/**
+ * TODO
+ * @param {Object} display - TODO
+ * @return {Boolean}
+ */
+function displayValidation(display) {
+  // TODO
+  return true;
+}
+
+/**
+ * TODO: jsdoc
+ */
+class InteractiveObjectServerBaseClass {
+  /**
+   * Create the standalone extension service object but does not start it.
+   * @constructor
+   * @param {ExtensionHelper} helper - An extension helper object for servicing
+   * various functionalities of the extension.
+   * @param {String} objectName - The name of this interactive object.
+   * @param {String} configFilePath - The path of the config file.
+   */
+  constructor(helper, objectName, configFilePath) {
+    this.helper = helper;
+
+    // load config
+    this.config = JSON.parse(fs.readFileSync(configFilePath));
+    if (!this.config.enabled) return;
+    this.objectName = objectName;
+
+    // input sanitization
+    for (const attr of ['initialPosition', 'display', 'FSM']) {
+      if (!(attr in this.config)) {
+        console.error(`missing attribute '${attr}' in ${configFilePath}`);
+        return;
+      }
+    }
+
+    if (!initialPositionValidation(this.config.initialPosition)) return;
+    if (!fsmValidation(this.config.FSM)) return;
+    if (!displayValidation(this.config.display)) return;
+
+    // TODO: store the states in database (or in `/run/small_data`)
+    this.dataStore = new Map(); // key: playerName, value: currentState
+  }
+
+  /**
+   * TODO
+   * @param {String} playerID - TODO
+   */
+  async fsmWalk(playerID) {
+    const fsm = this.config.FSM; // alias
+
+    // initialize if not exist
+    if (!this.dataStore.has(playerID)) this.dataStore.set(playerID, fsm.initialState);
+
+    this._fsmWalkExit = false;
+    while (!this._fsmWalkExit) {
+      const currState = fsm.states[this.dataStore.get(playerID)];
+      if (typeof currState === 'undefined') {
+        this.dataStore.set(playerID, fsm.initialState);
```

Comment:
Optional: Consider doing additional initialization and add a warning here. This is so that we can differentiate between players who are not yet initialized and when there's a misconfiguration in fsm.